Add episode 24 (#245)
trinodb/trino.io: _episodes/24.md

   applications, scaling the applications as requests grow and diminish, 
   redeploying crashed application, and networking the applications together.
   In summary, all of these activities can be considered container orchestration
   and this is exactly what Kubernetes solves!
   
   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/load-balancer.jpeg"/><br/>
    Source: https://www.slideshare.net/devopsdaysaustin/continuously-delivering-microservices-in-kubernetes-using-jenkins<br/>
    Here we hae two services that each sit behind a load balancer provided and mapped by the Kuberenets cluster.
   </p>

1. **Kubernetes components and architecture**:
   - Node - The physical machine or VM running in the kubernetes cluster. 
   - Cluster - a set of nodes connected to the same control plane node.
   - POD - single instance of an application, the smallest object in kubernetes.

   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/components-of-kubernetes.svg"/><br/>
    Source: https://kubernetes.io/docs/concepts/overview/components/
   </p>

>    Kubernetes control plane consists of:

Reviewer: same for other lists below

application to run on Kubernetes. However, at the end of the day we rely on 
databases that must exist outside, so it is only so useful.

## PR of the week: PR 11 Merge contributor version of k8s charts with the community version

This weeks [PR of the week](https://github.com/trinodb/charts/pull/11) comes 
from a different repo under the trinodb org, [trinodb/charts](https://github.com/trinodb/charts).
This PR contains the merging from contributer [Valeriano Manassero](https://github.com/valeriano-manassero). 

Valerino maintains a [very useful helm chart](https://github.com/valeriano-manassero/helm-charts/tree/main/valeriano-manassero/trino), 
that started before the Trino org had defined our own community chart. This pull
request effectively is trying to merge some useful features Valeriano added to 
his Trino helm chart so that it can be maintained in the community version.

Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>

It hasn't been merged yet but we are really looking forward to seeing this get
merged in. Thanks Valeriano!

## Demo: Running the Trino charts with kubectl

> For this weeks Demo, you will need to install [kubectl](https://kubernetes.io/docs/tasks/tools/),

Reviewer: demo, you need to

[minikube](https://minikube.sigs.k8s.io/docs/start/) using the [docker driver](https://minikube.sigs.k8s.io/docs/drivers/docker/),

> and [helm](https://helm.sh/docs/intro/install/). You will also be referencing 

Reviewer: Reword to avoid future tense

the ArtifactHub URL for this.

https://artifacthub.io/packages/helm/trino/trino

> ```

Reviewer: Break all this below up into multiple code snippets and text

    Here we hae two services that each sit behind a load balancer provided and mapped by the Kuberenets cluster.
   </p>

1. **Kubernetes components and architecture**:

   - Node - The physical machine or VM running a kubelet and container runtime.
   - Control Plane - The container orchestration layer that exposes the API and 
   interfaces to define, deploy, and manage the lifecycle of containers.
   - Cluster - a set of nodes connected to the same control plane.
   - Pod - single instance of an application, the smallest object in kubernetes.

   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/components-of-kubernetes.svg"/><br/>
    Source: https://kubernetes.io/docs/concepts/overview/components/
   </p>

Kubernetes control plane components:

 - API server that nodes connect with and is the front end for users and 
 administrators of the cluster.
 - etcd keystore is a distributed store containing all data used to manage 
 the cluster
 - Scheduler that distributes work across nodes and assigns newly created 
 containers to nodes
 - Controllers that are the brain behind orchestration and monitors for 
 nodes going down etc...
     

> Kubernetes worker nodes components:

Reviewer 2: ```suggestion
Kubernetes worker node components:
```

>  - container runtime - is the underlying runtime used to manage containers 

Reviewer 2: for consistency with other bullets below

```suggestion
 - container runtime - underlying runtime used to manage containers 
```

>  - kubelet - agent that checks the health and manages the pods running on the node based on the desired state provided in the PodSpec

Reviewer 2: I don't think PodSpec has been introduced before this point? No idea if it matters or not.

Author: If anything hasn't been covered we point to the kubernetes documentation

 - kube-proxy - network proxy that maintains network rules applied to nodes and allows network access between Pods in a cluster

   You can scale up multiple pods on a single node until the node has no more 
   resources, at which time a new node needs to be added and pod instances are 
   distributed between the nodes. 

> 1. **So how does this relate to Trino?**

Reviewer 2: Should be a heading maybe - gets kind of lost in the final page.

![image](https://user-images.githubusercontent.com/10473142/133820117-c7ef63a9-ce4e-4960-83c0-7d3270d456d4.png)

 - Out of the box, Kubernetes can do these key things for Trino.

>     - Simple scale up and down (manually tell k8s to startup or kill).

Reviewer 2: WDYT?

```suggestion
    - Simple scale up and down (manually tell k8s to start or kill Trino pods).
```

    - Kubernetes supports failover, meaning that your workers will restart if they die.
 - Advanced jobs that could exist but not currently in open source.
    - Auto-scaling via the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) 
    and custom metrics.
    - Graceful Shutdowns are hooks that you can add into your cluster that wait 
    to shut down to avoid a failed call to a node that already shut down.
     <p align="center">
        <img align="center" width="75%" height="100%" src="/assets/episode/24/kubernetes-shutdown.svg"/><br/>
        Source: https://learnk8s.io/graceful-shutdown
     </p>
     <p align="center">
        <img align="center" width="75%" height="100%" src="/assets/episode/24/graceful-shutdown.svg"/><br/>
        Source: https://learnk8s.io/graceful-shutdown
     </p>
     

> 1. **What the heck are helm charts then?**

Reviewer 2: Maybe a heading?

Author: Te be consistent with previous i'll keep this the same

- Helm is package manager for Kubernetes
- Removes the need for managing lots of Kubernetes related yaml files
- Best way to deploy apps to Kubernetes
- Charts are available for many different applications
- Helm chart for Trino

## PR of the week: PR 11 Merge contributor version of k8s charts with the community version

This weeks [PR of the week](https://github.com/trinodb/charts/pull/11) comes 
from a different repo under the trinodb org, [trinodb/charts](https://github.com/trinodb/charts).
This PR contains the merging from contributor [Valeriano Manassero](https://github.com/valeriano-manassero). 

Valerino maintains a [very useful helm chart](https://github.com/valeriano-manassero/helm-charts/tree/main/valeriano-manassero/trino), 
that started before the Trino org had defined our own community chart. This pull
request effectively is trying to merge some useful features Valeriano added to 
his Trino helm chart so that it can be maintained in the community version.

> Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>

Reviewer 2: ```suggestion
Valeriano's Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>
```

          imagePullPolicy: IfNotPresent
          volumeMounts:
            - mountPath: /etc/trino
              name: config-volume
            - mountPath: /etc/trino/catalog
              name: catalog-volume
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /v1/info
              port: http
          readinessProbe:
            httpGet:
              path: /v1/info
              port: http
          resources:
            {}
EOF
```

Now check the state of the k8s cluster again.
```
kubectl get all
```

Run the following command to expose the url and port to the localhost system.
```
minikube service tcb-trino --url
```

Clean up all the resources.
```
kubectl delete pod --all
kubectl delete replicaset --all
kubectl delete service tcb-trino
kubectl delete deployment tcb-trino-coordinator
kubectl delete configmap --all
```

Now you can run the same demo using the helm chart which includes all of these
templates out-of-the-box. First add the trino helm chart, check the templates
that are produced by helm, and run the install.
```

> # HELM DEMO

Reviewer 2: Is this heading suposed to be inside the code block or an actual heading?

Author: That's okay this was just to split it in the code portion of the demo but wasn't intended for markdown.

---
layout: episode
title:  "24: Trinetes I: Trino on Kubernetes"
date: 2021-08-19
tags: kubernetes trino containers
youtube_id: "BKrgnRuDVDw"
wistia_id: "vycp65bv7g"
sections: 
   - title: "Concept of the week"
     desc: "K8s architecture: Containers, Pods, and kubelets"
     time: 867
   - title: "PR of the week"
     desc: "PR 11 Merge contributor version of k8s charts with the community version"
     time: 3320
   - title: "Demo"
     desc: "Running the Trino charts with kubectl"
     time: 3462
---

This is the first episode in a series where we cover the basics and just enough
advanced Kubernetes features and information to understand how to deploy Trino 
on Kubernetes.

## Concept of the week: K8s architecture: Containers, Pods, and kubelets

For this concept of the week, we want to provide you a minimalistic overview of

> what you need to know for Kubernetes (also abbreviated k8s) to deploy a simple 

Reviewer: to know about Kubernetes to deploy Trino to a cluster.

Kubernetes cluster. 

1. **Why Kubernetes?** Kubernetes is a container orchestration platform that allows
   you to indicate how to manage containers declaritively using yaml 
   configuration files. This definition can be tricky to understand if you don't
   have proper context. To make sure nobody is left behind, it is useful to 
   cover what containers are.
    
   - The traditional way to deploy an application, is to take the compiled 
   binary of that application and run it directly on computer hardware that has
   an operating system to run the application on it. This works, but had a lot
   of dependency on the underlying hardware an operating system to be functional

>    and required multiple applications to share the same resources. If one of the

Reviewer: requires

>    applications fails and cause any of the shared resources to crash, it could

Reviewer: and causes

   affect all applications to fail on that machine. 
    
   - To remove these dependencies, engineers created virtual machines (VMs) by 

>    using a VM manager called the hypervisor that could run virtual environments 

Reviewer: that emulate hardware environments to host other operating systems.

   that hosted other operating systems. This was a big step forward as now each 

>    application could be isolated, but it came at a great cost. Each virtual machine

Reviewer: This is  ... can be isolated, but it comes at a

>    hosted an entire operating system and was resource intensive and slow.

Reviewer: hosts. .. and is

>    - Containers are the final type of deployment. Containers enable a logical

Reviewer: are the newest type

   isolation of resources while still physically running on shared resources. 
   All resources created in the hardware and operating systems exist on the host
   system. The isolation restricts any interference from other processes. 
   Containers achieve the goals of virtualization without sacrificing much 
   performance or efficiency. 
   

   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/container-evolution.svg"/><br/>
    Source: https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/
   </p>
   
   - Containerization simplified a trend in service architecture called 
   microservices. Microservices deploy loosely coupled and modular applications
   rather than all-encompassing monolithic applications. With containers, these

>    applications can be deployed and scale up quickly across various virtual and

Reviewer: and scaled up

   physical machines without affecting applications on the same machine. This

>    was great, but new problems emerged, such as, monitoring the health of

Reviewer: is great, but results in new complexities. Example are the need for new approaches to ..

   applications, scaling the applications as requests grow and diminish, 
   redeploying crashed application, and networking the applications together.
   In summary, all of these activities can be considered container orchestration
   and this is exactly what Kubernetes solves!
   
   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/load-balancer.jpeg"/><br/>
    Source: https://www.slideshare.net/devopsdaysaustin/continuously-delivering-microservices-in-kubernetes-using-jenkins<br/>
    Here we hae two services that each sit behind a load balancer provided and mapped by the Kuberenets cluster.
   </p>

1. **Kubernetes components and architecture**:

>    - Node - The physical machine or VM running a kubelet and container runtime. 

Reviewer: need to define what kubelet is before this  .. or avoid using it here

>    - Cluster - a set of nodes connected to the same control plane node.

Reviewer: what is control plan?

   - Pod - single instance of an application, the smallest object in kubernetes.

   <p align="center">
    <img align="center" width="75%" height="100%" src="/assets/episode/24/components-of-kubernetes.svg"/><br/>
    Source: https://kubernetes.io/docs/concepts/overview/components/
   </p>

   Kubernetes control plane components:

>      - API server that kubelets connect with and is the front end for 

Reviewer: empty line above list and dont indent the list items.

Reviewer: also all items in this list probably should be full sentences.

     users and administrators of the cluster.
     - etcd keystore is a distributed store containing all data used to manage 
     the cluster
     - Scheduler that distributes work across nodes and assigns newly created 
     containers to nodes
     - Controllers that are the brain behind orchestration and monitors for 
     nodes going down etc...
     

>    Kubernetes worker nodes components:

Reviewer: empty line before list and dont indent items

     - container runtime - is the underlying runtime used to manage containers 
     - kubelet - agent that checks the health and manages the pods running on the node based on the desired state provided in the PodSpec
     - kube-proxy - network proxy that maintains network rules applied to nodes and allows network access between Pods in a cluster
 

>    You can scale up multiple PODS on a single node until the node has no more 

Reviewer: pods

>    resources, at which time a new node will need to be added and pod instances will

Reviewer: needs to be added .. are distrbuted

   be distributed between the nodes. 

1. **So how does this relate to Trino?** 

>      - Why not just use docker compose? Docker-compose is a tool that creates

Reviewer: Indent left

Reviewer: also make this simpler .. or even remove docker-compose altogether ... 

If you want to keep it just use something like

- Why not just use docker-compose? Docker compose is only suitable for managing multi containers on one host. It has no orchestration features and does not support multiple hosts.

     multiple containers on a single host, where kubernetes is a platform that
     allows complex distributed workloads and self healing applications across
     nodes.

>      - Out of the box for any Kubernetes cluster can do these key things for Trino.

Reviewer: ? what does this sentence mean?

        - Simple scale up and down (manually tell k8s to startup or kill).

>         - Kubernetes supports failover, meaning that your workers will self heal if they die.

Reviewer: dont talk about self heal. .. Just use "restart"

     - Advanced jobs that could exist but not currently in open source.
        - Auto-scaling via the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) 
        and custom metrics.

>         - Graceful Shutdowns are hooks that you can add into your cluster that

Reviewer: something is missing here?

         <p align="center">
            <img align="center" width="75%" height="100%" src="/assets/episode/24/kubernetes-shutdown.svg"/><br/>
            Source: https://learnk8s.io/graceful-shutdown
         </p>
         <p align="center">
            <img align="center" width="75%" height="100%" src="/assets/episode/24/graceful-shutdown.svg"/><br/>
            Source: https://learnk8s.io/graceful-shutdown
         </p>

> 1. **What the heck are helm charts then?**

Reviewer: Empty line after title

> According to the helm site, Helm is the best way to find, share, and use 

Reviewer: Rejig this paragraph logically

- Helm is package manager for K8s
- Removes the need for managing lots of k8s related yaml files
- Best way to deploy apps to k8s
- Charts are available for many different applications
- Helm chart for Trino

software built for Kubernetes. What that comes down to is, you have to create 
and manage a lot of configurations before even starting kubernetes. This 
requires a lot of messing yml files and is tedious and time-consuming. In
essence, Helm is package management (packages all the complex configuration) 
with an opinionated default view of how a specific application gets deployed on
Kubernetes.
 
Fun Fact: Because Trino doesn't store data it becomes a good candidate 
application to run on Kubernetes. However, at the end of the day we rely on 
databases that must exist outside, so it is only so useful.

## PR of the week: PR 11 Merge contributor version of k8s charts with the community version

This weeks [PR of the week](https://github.com/trinodb/charts/pull/11) comes 
from a different repo under the trinodb org, [trinodb/charts](https://github.com/trinodb/charts).

> This PR contains the merging from contributer [Valeriano Manassero](https://github.com/valeriano-manassero). 

Reviewer: contributor

Valerino maintains a [very useful helm chart](https://github.com/valeriano-manassero/helm-charts/tree/main/valeriano-manassero/trino), 
that started before the Trino org had defined our own community chart. This pull
request effectively is trying to merge some useful features Valeriano added to 
his Trino helm chart so that it can be maintained in the community version.

Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>

It hasn't been merged yet but we are really looking forward to seeing this get
merged in. Thanks Valeriano!

## Demo: Running the Trino charts with kubectl

> For this weeks demo, you will need to install [kubectl](https://kubernetes.io/docs/tasks/tools/),

Reviewer: you need to
